specification/hugr.md

```diff
@@ -1168,44 +1168,78 @@ The new hugr is then derived as follows:
 
-  - a bijection μ<sub>out</sub> between out(G) and the set of output
-    ports of nodes in R whose target is not in R;
+Call two nodes $a, b \in \Gamma$ _separated_ if $a \notin \\{b\\}^\*$ and
+$b \notin \\{a\\}^\*$ (i.e. there is no hierarchy relation between them).
 
-  - disjoint subsets Init and Term of top(G);
+We define two structures to specify edge insertion.
 
-The new hugr is then derived by:
+A `NewInpEdgeSpec` specifies an edge inserted from an existing node to a (given) new node.
+It contains the following fields:
 
-1.  adding the new nodes from G;
+  - `EdgeKind`: may be `Value`, `Order`, `Static` or `ControlFlow`.
+  - `SrcNode`: the source node of the new edge.
+  - `SrcPort`: (for `Value` and `Static` edges only) the source port.
+  - `ToPos`: (for `Value` and `Static` edges only) the desired position among
+    the incoming ports to the new node.
 
-2.  connecting the ports according to the bijections μ<sub>inp</sub> and
-    μ<sub>out</sub>;
+A `NewOutEdgeSpec` specifies an edge inserted from a new node to an existing node.
+It contains the following fields:
 
-3.  for each node n in top(G), adding a hierarchy edge from t(n) to n,
-    placing n in the first position among children of t(n) if n is in
-    Init and in the second position if n is in Term;
+  - `EdgeKind`: may be `Value`, `Order`, `Static` or `ControlFlow`.
+  - `TgtNode`: the target node of the new edge.
+  - `TgtPort`: (for `Value` and `Static` edges only) the target port (whose
+    existing incoming edge will be removed).
+  - `FromPos` (for `Value`, `Static` and `ControlFlow` edges only) the desired
+    position among the outgoing ports from the new node.
 
-4.  for each node n in bot(G), and for each child m of b(n), adding a
-    hierarchy edge from n to m (replacing m’s existing parent edge)
+The `Replace` method takes as input:
 
-5.  removing all nodes in R
+  - a set $S$ of mutually-separated nodes in $\Gamma$;
+  - a partial hugr $G$;
+  - a map $T : \mathrm{top}(G) \to \Gamma \setminus S^*$ whose image consists of container nodes;
+  - a map $B : \mathrm{bot}(G) \to S^\*$ whose image consists of container nodes, such that $B(x)$
+    is separated from $B(y)$ unless $x = y$. Let $X$ be the set of children
+    of nodes $B(y)$ for $y \in \mathrm{bot}(G)$, and $R = S^\* \setminus X^\*$.
+  - a list $\mu\_\textrm{inp}$ of pairs $(n, \sigma\_\mathrm{inp})$ where $n$ is a node in $G$ and $\sigma\_\mathrm{inp}$ is a `NewInpEdgeSpec`;
+  - a list $\mu\_\textrm{out}$ of pairs $(n, \sigma\_\mathrm{out})$ where $n$ is a node in $G$ and $\sigma\_\mathrm{out}$ is a `NewOutEdgeSpec`.
 
-6.  If any edges inserted in step 2 are inter-graph (i.e DFG.
-    non-sibling), inserting any `Order` edges required to validate them.
+The new hugr is then derived as follows:
+
+1.  Make a copy in $\Gamma$ of all the nodes in $G$, and all edges between them.
+2.  For each $(n, \sigma\_\mathrm{inp}) \in \mu\_\textrm{inp}$, insert a new edge going into the new
```

Comment:
I wonder if it'd be easier for the spec (not the implementation, but the impl could do something differently) to say, "where these edges are from ports that currently have edges to nodes in R, the existing edges are replaced". Otherwise we are talking about some rather weird intermediates where the number of edges etc. don't line up. (That is to say - the spec doesn't have to talk about how we get from the start to the end, just what the end looks like)